Multi horizon

lisazeyen opened the pull request:
Closes # (if applicable).

## Changes proposed in this Pull Request
This PR should include perfect foresight for pathway optimisation.

Open ToDos:

- [x] add extra_functionalities for perfect foresight
- [ ] min_part_load for methanolisation for shipping and Fischer-Tropsch leads to infeasibilities
- [x] must run condition for existing gas boilers to avoid massive back up energy, add option to retrofit them to h2 boilers 
- [x] transmission expansion
- [x] carbon budget constraint
- [x] option for max growth for certain carriers

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`.
- [ ] Changes in configuration options are also documented in `doc/configtables/*.csv`.
- [x] A release note `doc/release_notes.rst` is added.

lisazeyen commented:
@FabianHofmann Some points still can be improved for the multi-decade optimisation, but I would be happy to pull a first beta version into the master branch and then work on the single features in separate smaller PR.

lisazeyen commented:
@FabianHofmann CI should run through once https://github.com/PyPSA/PyPSA/pull/721 is in. There is a problem if the optimisation is multi-decade and the list of extendable lines or DC links is empty when selecting the active assets of an empty list.

lisazeyen commented:
@FabianHofmann, what do you think, can we merge?

FabianHofmann commented on scripts/plot_summary.py:
```diff
@@ -379,18 +385,24 @@ def historical_emissions(countries):
     e["indirect"] = "ind_CO2 - Indirect CO2"
-    e["total wL"] = "Total (with LULUCF)"
-    e["total woL"] = "Total (without LULUCF)"
+    # e["total wL"] = "Total (with LULUCF)"
+    # e["total woL"] = "Total (without LULUCF)"
```
this is intended?

lisazeyen replied:
we can remove it for now. They changed the format in a newer version of the input data and then these two will be needed but I can remove it for now. We can update the data (which will effect also other scripts) in a separate PR.

FabianHofmann commented on scripts/solve_network.py:
```diff
@@ -47,6 +49,69 @@ def add_land_use_constraint(n, planning_horizons, config):
+    # p_nom_max = compress_series(p_nom_max)
+    # adjust name to fit syntax of nominal constraint per bus
+    df = p_nom_max.reset_index()
+    df["name"] = df.apply(
+        lambda row: f"nom_max_{row['carrier']}"
+        + (f"_{row['build_year']}" if row["build_year"] is not None else ""),
+        axis=1,
+    )
```
Why did you use the `n.buses["nom_max_{carrier}"]` scheme and not the `n.global_constraints`? I thought the latter is more flexible with the build_years.

lisazeyen replied:
I think it can be formulated in that way with a [max_relative_growth](https://github.com/PyPSA/PyPSA/blob/4514190afca73c8385b8fc50bbd9ad129b7d1c3c/pypsa/optimization/global_constraints.py#L170) so that the growth rate does not depend on an absolute number but on the installations in the previous period. We could think about adding an additional global constraint type for absolute growth numbers per investment period but I think we don't have this yet right?

FabianHofmann commented on scripts/solve_network.py:
```diff
@@ -130,16 +194,146 @@ def add_co2_sequestration_limit(n, limit=200):
         break
 
-    n.add(
+    if config["foresight"] == "perfect":
```
```suggestion
    if not n.investment_periods.empty:
```
This pattern is more object related. Could perhaps be used elsewhere. Then the `config` argument can also be removed.

FabianHofmann commented on scripts/solve_network.py:
```diff
@@ -130,16 +194,146 @@ def add_co2_sequestration_limit(n, limit=200):
     )
 
 
+def add_carbon_constraint(n, snapshots):
```
Could you shortly point out why this one (and the one below) is necessary? It feels like this part of code should rather go into pypsa itself, but not sure...

lisazeyen replied:
Mhm, I think one of the main reasons I implemented it directly in PyPSA-Eur is that with this implementation it is solving faster because the constraint is just on the CO2 atmosphere store and not on all components which have emissions. But this is maybe rather PyPSA-Eur specific since I guess many people do not model carbon management.

FabianHofmann commented on config/config.perfect.yaml:
```diff
@@ -0,0 +1,991 @@
+# SPDX-FileCopyrightText: : 2017-2023 The PyPSA-Eur Authors
```
I think we agreed on making this containing relative changes only, or? In VScode it is easy to see the relative change with the `Diff: File` command.

FabianHofmann reviewed (commented):
I would be much in favour to merge it. Here some last remaining comments and the CI should run through. Otherwise, it looks great!

lisazeyen commented:
@FabianHofmann thanks for your feedack. CI is running through with the latest PyPSA Version, only the building of the documentation fails.  I addressed your points above
